### PR TITLE
[android-auto] Dark/Light theme switch support

### PR DESCRIPTION
**Pull request by AndrewShkrob**
Demo:

https://github.com/organicmaps/organicmaps/assets/10351358/495d6676-08aa-4646-9824-d5c09eca1e80

https://developer.android.com/docs/quality-guidelines/car-app-quality

Navigation Functionality | NF-8 | The app must draw a light-themed or dark-themed map when instructed to do so.
-- | -- | --

Starting from the 30-second mark, when Dark Mode is enabled in `Auto` mode, the theme is adjusted based on the host information.
You can examine this behavior by utilizing the "N" button on the keyboard. (https://developer.android.com/training/cars/testing/dhu#day-night)

It's important to note that modifying the theme in Android Auto does not impact the theme on the device screen, and vice versa.

### CODE REVIEW

**Comment by AndrewShkrob** [comment]
I've implemented a fix for another issue I came across. Could you kindly confirm if this fix resolves your problem?

I'm unable to reproduce the bug you mentioned. Could you provide a comprehensive step-by-step guide? Include details such as whether you started AA first or the app on the device, the order in which you switched between screens, and any other relevant steps in the process.

**Comment by Jean-BaptisteC** [comment]
How to reproduce:
- Start OM on AA
- Start OM to phone
- Switch to phone
- Go to settings and enable night mode
- Switch to AA without leave settings -> Night mode is enabled on AA

With latest apk, I have map in night mode in AA not black screen

**Comment by AndrewShkrob** [comment]
Fixed

**Review by Jean-BaptisteC — APPROVED** [review]
✅ Pixel 6 - Android 13
Themes settings on phone and AA are independent

**Comment by biodranik** [comment]
1. What are the reasons to make a separate AA setting?
2. Does it detect and react to the car's light sensor?
3. Are there system AA settings for Light/Dark mode like [in CarPlay](https://support.apple.com/en-gb/guide/iphone/iph6ade13329/ios)?

**Comment by AndrewShkrob** [comment]
> 1. What are the reasons to make a separate AA setting?
> 2. Does it detect and react to the car's light sensor?
> 3. Are there system AA settings for Light/Dark mode like [in CarPlay](https://support.apple.com/en-gb/guide/iphone/iph6ade13329/ios)?

1. Because of the google's requirement on how maps should appear in `Auto` mode. Within our application, `Auto` mode is exclusively applied during navigation.
2. As for the process by which the host determines whether to present dark or light maps, it's uncertain. This might be based solely on time or could also involve the implementation of a light sensor. In our code, we merely receive an `isDarkMode` flag.
3. Yes.
   Screenshot 2023-08-31 at 09 40 10
   Screenshot 2023-08-31 at 09 40 13

**Comment by AndrewShkrob** [comment]
In brief, I hold a different viewpoint.

> Users will be surprised to see OM on AA in a light theme despite they set Dark one in AA settings.

This is exactly what is happening in the mobile application right now. There is a system theme setting, and there's an `Auto` theme which works only in the navigation mode and it doesn't correlate with the mode selected in the system.

12 a.m.
Select dark mode in system settings
Select `Auto` mode in app settings
In common mode the app theme is Light
In navigation mode the app theme is still Light
WTF?

> After your change, there will be 3 places where users can change dark/light mode settings.

There will be only two places in the AA. System + app.
And only two places on the mobile device. System + app (with broken logic as described above)

I don't see any complications here. If the user doesn't care about theme mode - the app will use the system's one by default. If the user wants to see our maps always in night or light mode, he/she can force it in the app settings.
If you don't like it this way, then, probably, we should also remove this switch in the mobile app settings because I don't see any reason why AA and mobile app should behave differently in the similar situation. But removing this switch in the mobile app sounds weird, right? :)
Absolutely, if users find the theme switch in AA undesirable, we can easily eliminate this feature in upcoming iterations.

File: android/src/app/organicmaps/util/ThemeSwitcher.java [hunk 1]
```diff
@@ -134,6 +135,10 @@ private void changeMapStyle(@NonNull String newTheme, @NonNull String oldTheme)
 
   private void SetMapStyle(@Framework.MapStyle int style)
   {
+    // Because of the distinct behavior in auto theme, Android Auto employs its own mechanism for theme switching.
```
Comment (rtsisyk) [inline android/src/app/organicmaps/util/ThemeSwitcher.java:138]: Why do we need the distinct behavior for Android Auto? Existing ThemeSwitcher is smart enough and checks lat,lon for getting sunrise/sunset time. This logic is quite useful.

**Comment by rtsisyk** [comment]
> This is exactly what is happening in the mobile application right now. There is a system theme setting, and there's an `Auto` theme which works only in the navigation mode and it doesn't correlate with the mode selected in the system.

I agree with you that the current day/night mode switch is confusing. There is a ticket for that: https://github.com/organicmaps/organicmaps/issues/749

I would rather use existing theme logic in the main app and fix this setting later, rather than implementing the new logic. Alternatively, you can rely fully on system settings from Android Auto.

**Comment by biodranik** [comment]
> This is exactly what is happening in the mobile application right now. There is a system theme setting, and there's an Auto theme which works only in the navigation mode and it doesn't correlate with the mode selected in the system.

No, you did not understand me. My comment was about a different behavior. The mobile app has a different issue, that should be solved separately.

In short: it's better to derive and follow the system setting in AA, in the same way as it is done in CarPlay now, at least in Auto mode. It allows to automatically switch the theme when a user or light sensor changes it. And it simplifies the support.

From the UX point of view, there are two devices: a car unit, and a phone in a hand. And they have their own _system_ dark mode settings that may work differently in different scenarios.

**Comment by AndrewShkrob** [comment]
@rtsisyk PTAL

**Comment by biodranik** [comment]
AA OM implementation should use the main OM app's settings, and do not introduce yet one more setting, to avoid complications and for easier user support:
1. Light or Dark mode setting in the main app also forces Light or Dark in the OM's AA.
2. Auto setting in the main app means that OM's AA implementation should use AA system settings for dark or light mode. Then if the car switches it (in a tunnel), OM's AA implementation will also correctly switch the theme.

**Review by biodranik — APPROVED** [review]
AA OM implementation should use the main OM app's settings, and do not introduce yet one more setting, to avoid complications and for easier user support:
1. Light or Dark mode setting in the main app also forces Light or Dark in the OM's AA.
2. Auto setting in the main app means that OM's AA implementation should use AA system settings for dark or light mode. Then if the car switches it (in a tunnel), OM's AA implementation will also correctly switch the theme.

**Review by biodranik — CHANGES_REQUESTED** [review]
AA OM implementation should use the main OM app's settings, and do not introduce yet one more setting, to avoid complications and for easier user support:
1. Light or Dark mode setting in the main app also forces Light or Dark in the OM's AA.
2. Auto setting in the main app means that OM's AA implementation should use AA system settings for dark or light mode. Then if the car switches it (in a tunnel), OM's AA implementation will also correctly switch the theme.

**Comment by rtsisyk** [comment]
> AA OM implementation should use the main OM app's settings, and do not introduce yet one more setting, to avoid complications and for easier user support:
> 
>     1. Light or Dark mode setting in the main app also forces Light or Dark in the OM's AA.
>     2. Auto setting in the main app means that OM's AA implementation should use AA system settings for dark or light mode. 

Light or Dark mode setting in the main doesn't correlate with Light or Dark in AA. The phone and the car display are two independent devices. 

> Then if the car switches it (in a tunnel), OM's AA implementation will also correctly switch the theme.

"Auto" setting in OM's AA already uses system settings from AA.

P.S. lint fails on LocationHelper after rebase.

**Comment by rtsisyk** [comment]
To summarize, the setting in the main app has nothing in common with dark/light mode in OM's AA app. At the current 
Currently, we have just two choices to select from:

1. Keep the proposed implementation that adds a switch in OM's AA app settings with three options:
  - Auto (=follows AA settings)
  - Dark (=forces dark)
  - Light (=forces light).
2. Keep only Auto (=follows AA settings) mode and do not add a switch in OM's AA app.

Since (1) provides extra flexibility to the users and it is already implemented in this PR, why should we spend more time on removing this feature from the PR? It doesn't sound reasonable to me. Moreover, the main app also has this switch in the settings instead of forcing “auto” mode for everyone.

**Review by rtsisyk — APPROVED** [review]

**Pull request merged**